code/controllers/subsystem/hijack.dm

```diff
@@ -0,0 +1,371 @@
+			shuttle.cancel_evac()
+		return TRUE
+
+/// Opens the lifeboat doors and gets them ready to launch
+/datum/controller/subsystem/hijack/proc/activate_lifeboats()
+	for(var/obj/docking_port/stationary/lifeboat_dock/lifeboat_dock in GLOB.lifeboat_almayer_docks)
+		var/obj/docking_port/mobile/crashable/lifeboat/lifeboat = lifeboat_dock.get_docked()
+		if(lifeboat && lifeboat.available)
+			lifeboat.status = LIFEBOAT_ACTIVE
+			lifeboat_dock.open_dock()
+
+/// Turns off ability to manually take off lifeboats
+/datum/controller/subsystem/hijack/proc/deactivate_lifeboats()
+	for(var/obj/docking_port/stationary/lifeboat_dock/lifeboat_dock in GLOB.lifeboat_almayer_docks)
+		var/obj/docking_port/mobile/crashable/lifeboat/lifeboat = lifeboat_dock.get_docked()
+		if(lifeboat && lifeboat.available)
+			lifeboat.status = LIFEBOAT_INACTIVE
+
+
+/// Once refueling is done, marines can optionally hold SD for a time for a stalemate instead of a xeno minor
+/datum/controller/subsystem/hijack/proc/unlock_self_destruct()
+	sd_time_remaining = sd_max_time
+	sd_unlocked = TRUE
+	marine_announcement("Fuel reserves full. Manual detonation of fuel reserves by overloading the on-board fusion reactors now possible.", HIJACK_ANNOUNCE)
+
+/datum/controller/subsystem/hijack/proc/on_generator_overload(obj/structure/machinery/power/fusion_engine/source, new_overloading)
+	SIGNAL_HANDLER
+
+	adjust_generator_overload_count(new_overloading ? 1 : -1)
+
+/datum/controller/subsystem/hijack/proc/adjust_generator_overload_count(amount = 1)
+	var/generator_overload_percent = round(overloaded_generators / maximum_overload_generators, 0.01)
+	var/old_required_time = sd_min_time + ((1 - generator_overload_percent) * (sd_max_time - sd_min_time))
+	var/percent_completion_remaining = sd_time_remaining / old_required_time
+	overloaded_generators = clamp(overloaded_generators + amount, 0, maximum_overload_generators)
+	generator_overload_percent = round(overloaded_generators / maximum_overload_generators, 0.01)
+	var/new_required_time = sd_min_time + ((1 - generator_overload_percent) * (sd_max_time - sd_min_time))
+	sd_time_remaining = percent_completion_remaining * new_required_time
+
+
+/datum/controller/subsystem/hijack/proc/detonate_sd()
+	set waitfor = FALSE
+	sd_detonated = TRUE
+	marine_announcement("ALERT: Fusion reactors dangerously overloaded. Runaway meltdown in reactor core imminent.")
+	sleep(5 SECONDS)
+
+	var/sound_picked = pick('sound/theme/nuclear_detonation1.ogg','sound/theme/nuclear_detonation2.ogg')
+	for(var/client/player as anything in GLOB.clients)
+		playsound_client(player, sound_picked, 90)
+
+	var/list/alive_mobs = list() //Everyone who will be destroyed on the zlevel(s).
+	var/list/dead_mobs = list() //Everyone who only needs to see the cinematic.
+	for(var/mob/current_mob as anything in GLOB.mob_list) //This only does something cool for the people about to die, but should prove pretty interesting.
+		if(!current_mob?.loc)
```

Comment:
having a loc doesn't guarantee having a turf z - see the same bug fixed in #4659